File: chapter_2_linked_lists/js/2.8-loopDetection.js
const { SinglyLinkedList } = require("./singlyLinkedList")

class SinglyLinkedList_2_8 extends SinglyLinkedList {
  /**
   * Determines if linked list is circular.
   * O(n) solution.
   * @returns   {boolean}
   */
  isCircular() {
    let hare = this.head
    let tortoise = this.head
    while (hare && hare.next) {
      hare = hare.next.next
      tortoise = tortoise.next
      if (hare === tortoise) return true
    }
    return false
  }
}

const sll = new SinglyLinkedList_2_8()
let s3 = sll.addNode(3)
sll.addNode(1)
sll.addNode(5)
sll.addNode(9)
sll.addNode(7)
let s2 = sll.addNode(2)
sll.addNode(8)
sll.printList() // 8 -> 2 -> 7 -> 9 -> 5 -> 1 -> 3
console.log(sll.isCircular()) // false

/* Making this circular. */
s3.next = s2
console.log(sll.isCircular()) // true

File: chapter_2_linked_lists/python/2.8-loop_detection.py
SinglyLinkedList = __import__('singly_linked_list').SinglyLinkedList


class SinglyLinkedList_2_8(SinglyLinkedList):

    def is_circular(self):
        """
        Determines if linked list is circular
        O(n) solution
        """
        hare = tortoise = self.head
        while hare and hare.next:
            hare = hare.next.next
            tortoise = tortoise.next
            if hare == tortoise:
                return True
        return False


sll = SinglyLinkedList_2_8()
sll.add_to_front(1)
sll.add_to_front(2)
sll.add_to_front(3)
n4 = sll.add_to_front(4)
sll.add_to_front(3)
sll.add_to_front(2)
sll.add_to_front(1)
sll.print_linked_list()  # 1-> 2-> 3-> 4-> 3-> 2-> 1
print(sll.is_circular())  # False

# making this circular
curr = sll.head
while curr.next:
    curr = curr.next
curr.next = n4

print(sll.is_circular())  # True
